src/Sulu/Component/Content/Document/Syncronization/DocumentRegistrator.php

```diff
@@ -0,0 +1,214 @@
+        $locale = $ddmInspector->getLocale($document);
+        $pdmRegistry->registerDocument(
+            $document,
+            $node,
+            $locale
+        );
+    }
+
+    /**
+     * If possible, resolve the UUID of the node in the PDM corresponding to
+     * the DDM node.
+     *
+     * If the UUID does not exist, we check to see if the path exsits.
+     * if neither the path or UUID exist, then the PDM should create a new
+     * document and we ensure that the PARENT path exists and if it doesn't
+     * we syncronize the ancestor nodes from the DDM.
+     *
+     * We return FALSE in the case that a new document should be created.
+     *
+     * In the case the UUID does not exist we assume that in a valid system
+     * that path will also NOT EXIST. If the path does exist, then it means that
+     * the corresponding PHPCR nodes were created independently of each
+     * other and bypassed the syncrhonization system and we throw an exception.
+     *
+     * @throws \RuntimeException If the UUID could not be resolved and it would be
+     *                           invalid to implicitly allow the node to be created.
+     */
+    private function resolvePDMUUID($object)
+    {
+        $pdmNodeManager = $this->publishManager->getNodeManager();
+        $ddmInspector = $this->defaultManager->getInspector();
+        $uuid = $ddmInspector->getUUid($object);
+
+        if (true === $pdmNodeManager->has($uuid)) {
+            return $uuid;
+        }
+
+        $path = $ddmInspector->getPath($object);
+
+        if (false === $pdmNodeManager->has($path)) {
+
+            // if the parent path also does not exist in the PDM then we need
+            // to create the parent path using the same UUIDs that are used in
+            // the DDM.
+            $parentPath = PathHelper::getParentPath($path);
+            if (false === $pdmNodeManager->has($parentPath)) {
+                $this->syncPDMPath($parentPath);
+
+                return false;
+            }
+
+            return false;
+        }
+
+        throw new \RuntimeException(sprintf(
+            'Publish document manager already has a node at path "%s" but ' .
+            'incoming UUID `%s` does not match existing UUID: "%s".',
+            $path, $uuid, $pdmNodeManager->find($path)->getIdentifier()
+        ));
+    }
+
+    /**
+     * Sync the given path from the DDM to the PDM, preserving
+     * the UUIDs.
+     *
+     * @param string $path
+     */
+    private function syncPDMPath($path)
+    {
+        $ddmNodeManager = $this->defaultManager->getNodeManager();
+        $pdmNodeManager = $this->publishManager->getNodeManager();
+        $segments = explode('/', $path);
+        $stack = [];
+
+        foreach ($segments as $segment) {
+            $stack[] = $segment;
+
+            if ($segment == '') {
+                continue;
+            }
+
+            $path = implode('/', $stack) ?: '/';
```

Comment:
Any reason not just to use string concatenation? So the path gets a little bit longer with every iteration?